File: solutions/cpp/queue_from_stacks.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <limits>
#include <stack>
#include <stdexcept>

using std::cout;
using std::endl;
using std::exception;
using std::length_error;
using std::numeric_limits;
using std::stack;

// @include
class Queue {
 public:
  void Enqueue(int x) { enqueue_.emplace(x); }

  int Dequeue() {
    if (dequeue_.empty()) {
      // Transfers the elements in enqueue_ to dequeue_.
      while (!enqueue_.empty()) {
        dequeue_.emplace(enqueue_.top());
        enqueue_.pop();
      }
    }

    if (dequeue_.empty()) {  // dequeue_ is still empty!
      throw length_error("empty queue");
    }
    int ret = dequeue_.top();
    dequeue_.pop();
    return ret;
  }

 private:
  stack<int> enqueue_, dequeue_;
};
// @exclude

int main(int argc, char* argv[]) {
  Queue Q;
  Q.Enqueue(1);
  Q.Enqueue(2);
  assert(1 == Q.Dequeue());
  assert(2 == Q.Dequeue());
  Q.Enqueue(3);
  assert(3 == Q.Dequeue());
  try {
    Q.Dequeue();
    assert(false);
  } catch (const exception& e) {
    cout << e.what() << endl;
  }
  Q.Enqueue(-1);
  Q.Enqueue(123);
  Q.Enqueue(numeric_limits<int>::max());
  Q.Enqueue(numeric_limits<int>::min());
  Q.Enqueue(0);
  assert(-1 == Q.Dequeue());
  Q.Enqueue(0);
  assert(123 == Q.Dequeue());
  assert(numeric_limits<int>::max() == Q.Dequeue());
  assert(numeric_limits<int>::min() == Q.Dequeue());
  assert(0 == Q.Dequeue());
  assert(0 == Q.Dequeue());
  try {
    Q.Dequeue();
    assert(false);
  } catch (const exception& e) {
    cout << e.what() << endl;
  }
  return 0;
}

File: solutions/python/queue_from_stacks.py
# @include
class Queue:

    def __init__(self):
        self._enq, self._deq = [], []

    def enqueue(self, x):
        self._enq.append(x)

    def dequeue(self):
        if not self._deq:
            # Transfers the elements in _enq to _deq.
            while self._enq:
                self._deq.append(self._enq.pop())

        if not self._deq:  # _deq is still empty!
            raise IndexError('empty queue')
        return self._deq.pop()
# @exclude


def main():
    Q = Queue()
    Q.enqueue(1)
    Q.enqueue(2)
    assert 1 == Q.dequeue()  # 1
    assert 2 == Q.dequeue()  # 2
    Q.enqueue(3)
    assert 3 == Q.dequeue()  # 3
    try:
        Q.dequeue()
        assert False
    except IndexError as e:
        print(e)  # throw
    Q.enqueue(-1)
    Q.enqueue(123)
    Q.enqueue(2**64 - 1)
    Q.enqueue(-2**64)
    Q.enqueue(0)
    assert -1 == Q.dequeue()
    Q.enqueue(0)
    assert 123 == Q.dequeue()
    assert 2**64 - 1 == Q.dequeue()
    assert -2**64 == Q.dequeue()
    assert 0 == Q.dequeue()
    assert 0 == Q.dequeue()
    try:
        Q.dequeue()
        assert False
    except IndexError as e:
        print(e)  # throw


if __name__ == '__main__':
    main()
